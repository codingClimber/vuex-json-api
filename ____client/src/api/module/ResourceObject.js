export class ResourceObject {
  /**
   *
   * @param {Vuex.Store} store
   * @param {object} resultObject
   */
  constructor (store, resultObject) {
    this.store = store

    if (DataTransferItem.hasOwnProperty('relationships')) {
      resultObject.relationships = this.addRelationshipResolvers(resultObject.relationships)
    }

    return resultObject
  }

  /**
   *
   * @param {object} resultObjectRelationships
   */
  addRelationshipResolvers (resultObjectRelationships) {
    for (const relatedObjectType in resultObjectRelationships) {
      if (resultObjectRelationships.hasOwnProperty(relatedObjectType)) {
        resultObjectRelationships[relatedObjectType].get = this.resolveRelationshipForObjectType(relatedObjectType)
      }
    }

    return resultObjectRelationships
  }

  /**
   *
   * @param {String} relatedObjectType
   */
  resolveRelationshipForObjectType (relatedObjectType) {
    return new Proxy(() => {}, {
      apply (target, thisArg, argArray) {
        let isToManyRelationship = thisArg.data.constructor === Array

        let relationshipType = null
        let relationshipItemId = null

        if (isToManyRelationship) {
          relationshipType = thisArg.data[0].type
          relationshipItemId = thisArg.data[argArray[0]].id
        } else {
          relationshipType = thisArg.data.type
          relationshipItemId = thisArg.data.id
        }

        let relationshipModule = this.store.state[relationshipType]

        return relationshipModule.items[relationshipItemId]
      }
    })

    // TODO: if relationship is a toMany, it should also have an all() method or something like that
    // TODO: to get the collection of all related items
  }
}
