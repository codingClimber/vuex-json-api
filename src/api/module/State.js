/**
 * Return a new Object representing the initial state of a module
 *
 * The state is returned from inside a self-calling closure to
 * make absolutely sure we get pristine objects. This is a safety
 * measure to guard against unintended cross-module reference bindings.
 *
 * ## The fields
 *
 * **item/items**:
 *
 * This is the container for the normalized resource objects of the
 * module. If the module is not a collection, there will always only
 * be one item, thus it's singularized.
 *
 * **initial**:
 *
 * Once any data-changing mutation is applied to the state,
 * the store will automatically copy the affected resource object
 * into `initial`. This is always either empty or a map of `id => ResourceObject`
 *
 * **currentPage/totalPages**
 *
 * If the API responses contain pagination information in `meta`
 * this will be stored in these properties and subsequently
 * used/updated upon navigation
 *
 * @param {Boolean} isCollection
 */
function initialState (isCollection) {
  return (() => {
    if (isCollection) {
      return {
        loading: false,
        items: {},
        initial: {},
        currentPage: null,
        totalPages: null
      }
    } else {
      return {
        loading: false,
        item: {},
        initial: {}
      }
    }
  })()
}

/**
 * Check whether the api supports collections
 *
 * @param {object} apiMethods supported api methods (list, get, update, delete)
 */
function isCollection (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('list')
}

function allowsModification (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('update')
}

function allowsReplacement (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('replace')
}

function allowsCreation (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('create')
}

function allowsDeletion (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('delete')
}

export {
  allowsCreation,
  allowsDeletion,
  allowsModification,
  allowsReplacement,
  initialState,
  isCollection
}
