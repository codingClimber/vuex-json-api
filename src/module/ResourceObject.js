export class ResourceObject {
  /**
   *
   * @param {Vuex.Store} store
   * @param {object} resultObject
   */
  constructor (store, resultObject) {
    let obj = JSON.parse(JSON.stringify(resultObject.data))

    if (obj.hasOwnProperty('relationships')) {
      obj.relationships = this.addRelationshipResolvers(store, obj.id, obj.type, obj.relationships)
    }

    obj.hasRelationship = name => {
      return obj.hasOwnProperty('relationships') &&
        obj.relationships.hasOwnProperty(name) &&
        obj.relationships[name].data.length !== 0
    }

    obj.hasLoadedRelationship = name => {
      if (!obj.hasRelationship(name)) {
        return false
      }

      try {
        obj.relationships[name].get() || obj.relationships[name].list()
        return true
      } catch (e) {
        return false
      }
    }

    return obj
  }

  /**
   *
   * @param {Vuex.Store} store
   * @param {object} resultObjectRelationships
   */
  addRelationshipResolvers (store, currentObjectId, currentObjectType, resultObjectRelationships) {
    for (const relatedObjectType in resultObjectRelationships) {
      if (resultObjectRelationships.hasOwnProperty(relatedObjectType)) {
        const isToManyRelationship = resultObjectRelationships[relatedObjectType].data.constructor === Array

        let relatedObject = resultObjectRelationships[relatedObjectType]

        relatedObject.get = this.resolveToOneRelationship(store, relatedObject, isToManyRelationship)

        if (isToManyRelationship) {
          relatedObject.list = this.resolveToManyRelationship(store, relatedObject)
        }

        relatedObject.load = this.addLoaderMethod(store, currentObjectId, currentObjectType, relatedObject, isToManyRelationship)

        resultObjectRelationships[relatedObjectType] = relatedObject
      }
    }

    return resultObjectRelationships
  }

  addLoaderMethod (store, currentObjectId, currentObjectType, relatedObject, isToManyRelationship) {
    return new Proxy(() => { }, {
      apply (target, thisArg, argArray) {
        let relatedObjectType = (isToManyRelationship) ? relatedObject.data[0].type : relatedObject.data.type
        let relatedObjectNameForAction = relatedObjectType.getChar(0).toUpperCase() + relatedObjectType.slice(1)

        if (isToManyRelationship) {
          return store.dispatch(`${currentObjectType}/listRelated${relatedObjectNameForAction}`, { currentObjectId })
        }
      }
    })
  }

  /**
   *
   * @param {Vuex.Store} store
   * @param {Object} relatedObject
   */
  resolveToOneRelationship (store, relatedObject, isToManyRelationship) {
    return new Proxy(() => {}, {
      apply (target, thisArg, argArray) {
        let moduleName = relatedObject.data.type
        let relatedModule = store.state[moduleName]

        if (isToManyRelationship) {
          let [requestedId] = argArray

          try {
            return relatedModule.items[requestedId]
          } catch (e) {
            throw new Error(`Related object ${relatedObject.id} not found in ${moduleName}`)
          }
        }

        return relatedModule.items[relatedObject.data.id]
      }
    })
  }

  /**
   * @param {Vuex.Store} store
   * @param {Array} relatedObjects
   */
  resolveToManyRelationship (store, relatedObjects) {
    return new Proxy(() => {}, {
      apply (target, thisArg, argArray) {
        let moduleName = relatedObjects.data[0].type
        let relatedModule = store.state[moduleName]

        let relatedObjectIds = relatedObjects.data.map(obj => obj.id)

        let relatedItems = {}
        relatedObjectIds.forEach(id => {
          relatedItems[id] = relatedModule.items[id]
        })

        return relatedItems
      }
    })
  }
}
