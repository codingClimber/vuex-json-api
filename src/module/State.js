import { item } from './state/item'
import { collection } from './state/collection'

/**
 * Return a new Object representing the initial state of a module
 *
 * The state is returned from inside a self-calling closure to
 * make absolutely sure we get pristine objects. This is a safety
 * measure to guard against unintended cross-module reference bindings.
 *
 * @param {Boolean} isCollection
 */
function initialState (isCollection) {
  return (() => {
    if (isCollection) {
      return JSON.parse(JSON.stringify(collection))
    } else {
      return JSON.parse(JSON.stringify(item))
    }
  })()
}

/**
 * Check whether the api supports collections
 *
 * @param {object} apiMethods supported api methods (list, get, update, delete)
 */
function isCollection (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('list')
}

function allowsModification (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('update')
}

function allowsReplacement (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('replace')
}

function allowsCreation (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('create')
}

function allowsDeletion (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('delete')
}

function hasRelated (apiMethods) {
  return typeof apiMethods === 'object' && apiMethods.hasOwnProperty('related') && typeof apiMethods.related === 'object'
}

export {
  allowsCreation,
  allowsDeletion,
  allowsModification,
  allowsReplacement,
  hasRelated,
  initialState,
  isCollection
}
