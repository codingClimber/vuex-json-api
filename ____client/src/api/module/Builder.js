import { initialState, isCollection, allowsDeletion, allowsModification } from './State'
import { get as getAction } from './actions/Get'
import { set as setAction } from './actions/Set'
import { list as listAction } from './actions/List'
import { reset as resetMutation } from './mutations/Reset'
import { set as setMutation } from './mutations/Set'
import { remove as removeMutation } from './mutations/Delete'

/**
 * JsonApi-based module builder for Vuex
 *
 * This module builder will create a vuex module based on the assumption of
 * working with valid json api resources. Additionally, it can do a little
 * more magic if the API adds the following meta data:
 *
 * - included: a list of included types (This is used to hookup relationships
 *             without incurring too much of a computation penalty for tree
 *             traversal
 * - the proposed json api 1.1 pagination style meta attributes
 *   (-> https://jsonapi.org/format/1.1/#fetching-pagination)
 */
export class Builder {
  /**
   * Module Builder for Json:Api bound Vuex Modules
   *
   * @param {Vuex} store
   * @param {ResourcefulApi} api
   * @param {String} moduleName
   * @param {Array[String]} apiMethods
   */
  constructor (store, api, moduleName, apiMethods) {
    this.store = store
    this.api = api
    this.moduleName = moduleName
    this.apiMethods = apiMethods

    this.isCollection = isCollection(apiMethods)
  }

  /**
   * Build the module for this builder instance
   */
  build () {
    const storeModuleBuildTimer = 'api: build module ' + this.moduleName
    console.time(storeModuleBuildTimer)

    let module = {
      namespaced: true,
      state: initialState(this.isCollection)
    }

    module['actions'] = this.buildActions()
    module['mutations'] = this.buildMutations()

    console.timeEnd(storeModuleBuildTimer)
    return module
  }

  /**
   * Build the mutations
   */
  buildMutations () {
    let mutations = {
      reset: resetMutation(this.isCollection),
      set: setMutation(this.store, this.isCollection)
    }

    if (allowsDeletion(this.apiMethods)) {
      mutations['remove'] = removeMutation(this.isCollection)
    }

    return mutations
  }

  buildActions () {
    let actions = {
      get: getAction(this.api, this.moduleName)
    }

    if (this.isCollection) {
      actions['list'] = listAction(this.api, this.moduleName)
    }

    if (allowsModification(this.apiMethods)) {
      actions['set'] = setAction()
    }

    return actions
  }
}
