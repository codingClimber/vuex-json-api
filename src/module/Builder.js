import { initialState, isCollection, allowsDeletion, allowsModification, allowsCreation } from './State'
import { addGroupAction } from './actions/addGroupAction'
import { addGroupMutation } from './mutations/addGroupMutation'
import { createAction } from './actions/createAction'
import { getAction } from './actions/getAction'
import { getProperty } from './getters/getProperty'
import { listAction } from './actions/listAction'
import { listTypeAndId } from './getters/listTypeAndId'
import { resetItemsAction } from './actions/resetItemsAction'
import { removeMutation } from './mutations/removeMutation'
import { resetItemsMutation } from './mutations/resetItemsMutation'
import { restoreFromInitialAction } from './actions/restoreFromInitialAction'
import { saveAction } from './actions/saveAction'
import { setAction } from './actions/setAction'
import { setMutation } from './mutations/setMutation'
import { setPaginationMutation } from './mutations/setPaginationMutation'
import { startLoadingMutation, endLoadingMutation } from './mutations/loading'
import { updateMutation } from './mutations/updateMutation'

/**
 * JsonApi-based module builder for Vuex
 *
 * This module builder will create a vuex module based on the assumption of
 * working with valid json api resources.
 *
 * - the proposed json api 1.1 pagination style meta attributes
 *   (-> https://jsonapi.org/format/1.1/#fetching-pagination)
 */
export class Builder {
  /**
   * Module Builder for Json:Api bound Vuex Modules
   *
   * @param {Vuex} store
   * @param {ResourcefulApi} api
   * @param {String} moduleName
   * @param {Array} apiMethods
   * @param {Object} options additional objects for the module builder
   */
  constructor (store, api, moduleName, apiMethods, options = {}) {
    this.store = store
    this.api = api
    this.moduleName = moduleName
    this.apiMethods = apiMethods || {}

    this.isCollection = isCollection(apiMethods)

    // is this a standalone module with no outside connections?
    this.isStandalone = options.hasOwnProperty('standalone') && options.standalone

    if (this.isStandalone) {
      // standalone modules are always collections
      // because they will only be created if a list request
      // ends up having unknown includes
      // and as the law of known unknowns and unknown unknowns goes
      // we know that we may get this unknown other resource
      // but we don't know how many items of that resource there will be
      this.isCollection = true
    }
  }

  /**
   * Build the module for this builder instance
   */
  build () {
    const storeModuleBuildTimer = 'api: build module ' + this.moduleName
    console.time(storeModuleBuildTimer)

    let module = {
      namespaced: true,
      state () { return initialState(isCollection) }
    }

    module['mutations'] = this.buildMutations()

    if (!this.isStandalone) {
      module['actions'] = this.buildActions()
      module['getters'] = this.buildGetters()
    }

    console.timeEnd(storeModuleBuildTimer)
    return module
  }

  /**
   * Build the mutations
   */
  buildMutations () {
    let mutations = {
      resetItems: resetItemsMutation(this.isCollection),
      set: setMutation(this.store, this.isCollection),
      startLoading: startLoadingMutation,
      endLoading: endLoadingMutation,
      update: updateMutation
    }

    if (allowsDeletion(this.apiMethods)) {
      mutations['remove'] = removeMutation(this.isCollection)
    }

    if (this.isCollection) {
      mutations['setPagination'] = setPaginationMutation
      mutations['addGroup'] = addGroupMutation
    }

    return mutations
  }

  buildActions () {
    let actions = {
      get: getAction(this.api, this.moduleName),
      resetItems: resetItemsAction,
      restoreFromInitial: restoreFromInitialAction(this.moduleName, this.isCollection)
    }

    if (this.isCollection) {
      actions['list'] = listAction(this.api, this.moduleName)
      actions['addGroup'] = addGroupAction
    }

    if (allowsModification(this.apiMethods)) {
      actions['set'] = setAction
      actions['save'] = saveAction(this.api, this.isCollection, this.moduleName)
    }

    if (allowsCreation(this.apiMethods)) {
      actions['create'] = createAction(this.api, this.moduleName)
    }

    return actions
  }

  buildGetters () {
    let getters = {
      getProperty: getProperty,
      listTypeAndId: listTypeAndId
    }

    return getters
  }
}
