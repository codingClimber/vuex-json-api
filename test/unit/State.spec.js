//import Vuex from 'vuex'
import { initialState } from '../../src/module/State';

it('returns the default state for a non-collection module', () => {
  let state = initialState(false)
  expect(state).toMatchSnapshot()
})
