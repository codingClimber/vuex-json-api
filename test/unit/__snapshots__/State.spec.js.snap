// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns the default state for a non-collection module 1`] = `
Object {
  "initial": Object {},
  "item": Object {},
  "loading": false,
}
`;
